File: src/domain/evento-resumo.js
// ═══════════════════════════════════════════════════════════
// EventoResumo / DiaResumo — agregados pós-execução
// ═══════════════════════════════════════════════════════════
// Lê stints (modelo do app local) e calcula totais e melhores tempos por dia
// e por evento. Usado pelo card de evento (`/app`) e relatório pós-evento.
//
// Quando stint tem `lapsExecutados[]` (vindos do pipeline real de telemetria):
//   conta voltas válidas + min(tempoMs); quality = 'MEDIDO'.
//
// Quando stint só tem `voltasPlanejadas` (entrada manual, sem telemetria):
//   usa esse valor; quality = 'ESTIMADO' (tempo fica null).
//
// Mistura → quality = 'PARCIAL'. Sem stints → 'VAZIO'.
//
// Esquema canônico no domain:
//   StintLap = { id?, tempoMs: number, valida?: boolean (default true) }
//   DiaResumo = { voltas, melhorMs, melhorLapId, melhorStintId, quality, calculadoEm }
//   EventoResumo = { voltasTotal, melhorMsGlobal, melhorMsPorDia, quality, calculadoEm }

export const ResumoQuality = Object.freeze({
  MEDIDO:   'MEDIDO',
  PARCIAL:  'PARCIAL',
  ESTIMADO: 'ESTIMADO',
  VAZIO:    'VAZIO',
});

function _melhorDosLaps(laps) {
  let best = null;
  let bestId = null;
  for (const l of laps) {
    if (!l || l.valida === false) continue;
    if (typeof l.tempoMs !== 'number' || !isFinite(l.tempoMs) || l.tempoMs <= 0) continue;
    if (best == null || l.tempoMs < best) { best = l.tempoMs; bestId = l.id || null; }
  }
  return { tempoMs: best, lapId: bestId };
}

export function calcularStintResumo(stint) {
  if (!stint) return _vazioStint();
  const laps = Array.isArray(stint.lapsExecutados) ? stint.lapsExecutados : [];
  if (laps.length > 0) {
    const validas = laps.filter(l => l && l.valida !== false && typeof l.tempoMs === 'number' && l.tempoMs > 0);
    const best = _melhorDosLaps(validas);
    return {
      voltas: validas.length,
      voltasTotais: laps.length,
      melhorMs: best.tempoMs,
      melhorLapId: best.lapId,
      quality: ResumoQuality.MEDIDO,
    };
  }
  if (typeof stint.voltasPlanejadas === 'number' && stint.voltasPlanejadas > 0) {
    return {
      voltas: stint.voltasPlanejadas,
      voltasTotais: stint.voltasPlanejadas,
      melhorMs: null,
      melhorLapId: null,
      quality: ResumoQuality.ESTIMADO,
    };
  }
  return _vazioStint();
}

function _vazioStint() {
  return { voltas: 0, voltasTotais: 0, melhorMs: null, melhorLapId: null, quality: ResumoQuality.VAZIO };
}

export function calcularDiaResumo(stints) {
  if (!Array.isArray(stints) || !stints.length) {
    return { voltas: 0, melhorMs: null, melhorLapId: null, melhorStintId: null, quality: ResumoQuality.VAZIO, calculadoEm: Date.now() };
  }
  let voltas = 0;
  let melhorMs = null;
  let melhorLapId = null;
  let melhorStintId = null;
  let temMedido = false;
  let temEstimado = false;
  for (const s of stints) {
    const r = calcularStintResumo(s);
    voltas += r.voltas;
    if (r.melhorMs != null && (melhorMs == null || r.melhorMs < melhorMs)) {
      melhorMs = r.melhorMs;
      melhorLapId = r.melhorLapId;
      melhorStintId = s.id || null;
    }
    if (r.quality === ResumoQuality.MEDIDO) temMedido = true;
    else if (r.quality === ResumoQuality.ESTIMADO) temEstimado = true;
  }
  let quality = ResumoQuality.VAZIO;
  if (temMedido && temEstimado) quality = ResumoQuality.PARCIAL;
  else if (temMedido) quality = ResumoQuality.MEDIDO;
  else if (temEstimado) quality = ResumoQuality.ESTIMADO;
  return { voltas, melhorMs, melhorLapId, melhorStintId, quality, calculadoEm: Date.now() };
}

export function calcularEventoResumo(evento, stintsByDia) {
  if (!evento || !Array.isArray(evento.dias)) {
    return { voltasTotal: 0, melhorMsGlobal: null, melhorMsPorDia: {}, quality: ResumoQuality.VAZIO, calculadoEm: Date.now() };
  }
  let voltasTotal = 0;
  let melhorMsGlobal = null;
  const melhorMsPorDia = {};
  const qualidades = [];
  for (const d of evento.dias) {
    const stints = (stintsByDia || {})[d.id] || [];
    const dr = calcularDiaResumo(stints);
    voltasTotal += dr.voltas;
    melhorMsPorDia[d.id] = dr.melhorMs;
    if (dr.melhorMs != null && (melhorMsGlobal == null || dr.melhorMs < melhorMsGlobal)) {
      melhorMsGlobal = dr.melhorMs;
    }
    qualidades.push(dr.quality);
  }
  const naoVazias = qualidades.filter(q => q !== ResumoQuality.VAZIO);
  let quality = ResumoQuality.VAZIO;
  if (naoVazias.length) {
    if (naoVazias.every(q => q === ResumoQuality.MEDIDO)) quality = ResumoQuality.MEDIDO;
    else if (naoVazias.every(q => q === ResumoQuality.ESTIMADO)) quality = ResumoQuality.ESTIMADO;
    else quality = ResumoQuality.PARCIAL;
  }
  return { voltasTotal, melhorMsGlobal, melhorMsPorDia, quality, calculadoEm: Date.now() };
}

// Mutates: escreve `dia.resumo` em cada dia + `evento.resumo`.
// Idempotente — pode ser chamada N vezes sem efeito colateral além do recalc.
export function recalcularResumoEvento(evento) {
  if (!evento || !Array.isArray(evento.dias)) return null;
  const stintsByDia = evento.stintsByDia || {};
  for (const d of evento.dias) {
    const stints = stintsByDia[d.id] || [];
    d.resumo = calcularDiaResumo(stints);
  }
  evento.resumo = calcularEventoResumo(evento, stintsByDia);
  return evento.resumo;
}

// ─── Enriquecer com Laps reais do Dexie ────────────────────────────────────
// Quando o cockpit roda com track configurada, SessionRecorder grava Laps em
// `db.laps` ligados a uma `Sessions`. Esta função busca por carId+data, cola
// `lapsExecutados` em cada stint do dia (via match por carId), e recalcula.
// Sem laps reais, stint.lapsExecutados fica intocado e fallback ESTIMATED
// continua valendo. Chamar em bootApp após carregar Dexie + DB local.

export async function enriquecerEventosComLapsReais(eventos, dbDexie) {
  if (!eventos?.length || !dbDexie) return { enriquecidos: 0 };
  let enriquecidos = 0;
  for (const ev of eventos) {
    let mudou = false;
    for (const d of (ev.dias || [])) {
      const dataDia = d.data;
      if (!dataDia) continue;
      const stints = (ev.stintsByDia || {})[d.id] || [];
      if (!stints.length) continue;

      // Sessions iniciadas neste dia (dataInicio é setado por SessionMaster.claim;
      // se ainda nulo, usa criadoEm como proxy — sessão recém-criada na mesma data).
      const allSessions = await dbDexie.sessions.toArray();
      const sessionsDoDia = allSessions.filter(s => {
        const ts = s.dataInicio || (s.criadoEm ? new Date(s.criadoEm).toISOString() : null);
        if (!ts) return false;
        return String(ts).slice(0, 10) === dataDia;
      });
      if (!sessionsDoDia.length) continue;

      for (const stint of stints) {
        if (!stint.carId) continue;
        const sessionsDoCarro = sessionsDoDia.filter(s => s.carId === stint.carId);
        if (!sessionsDoCarro.length) continue;
        const sessionIds = sessionsDoCarro.map(s => s.id);
        const lapsAll = await dbDexie.laps.where('sessionId').anyOf(sessionIds).toArray();
        const lapsLimpos = lapsAll
          .filter(l => l.tempoMs != null && l.tempoMs > 0)
          .map(l => ({ id: l.id, tempoMs: l.tempoMs, valida: l.valida !== false }));
        if (lapsLimpos.length) {
          stint.lapsExecutados = lapsLimpos;
          stint.sessionIds = sessionIds;  // pra debug + futura UI
          mudou = true;
        }
      }
    }
    if (mudou) {
      recalcularResumoEvento(ev);
      enriquecidos++;
    }
  }
  return { enriquecidos };
}

File: ios/p1fast-core/Sources/P1FastCore/EventoResumo.swift
// ═══════════════════════════════════════════════════════════
// EventoResumo / DiaResumo — port Swift de
//   src/domain/evento-resumo.js (camada pura, sem IO)
// ═══════════════════════════════════════════════════════════
// Lê stints e calcula totais e melhores tempos por dia e por evento.
// Usado pelo card de evento e pelo relatório pós-evento.
//
// Quality:
//   MEDIDO    — tudo veio de telemetria real (`lapsExecutados`)
//   ESTIMADO  — tudo veio de `voltasPlanejadas` (sem telemetria)
//   PARCIAL   — mistura
//   VAZIO     — sem stints
//
// MS-6.6 (port 2026-05-04). A versão JS `enriquecerEventosComLapsReais`
// é IO Dexie — não portada. Em iOS o caller faz o equivalente via
// GRDB (lap_repository fetchByCarroDia).

import Foundation

public enum ResumoQuality: String, Sendable, Codable, CaseIterable {
    case medido    = "MEDIDO"
    case parcial   = "PARCIAL"
    case estimado  = "ESTIMADO"
    case vazio     = "VAZIO"
}

/// Lap executado durante o stint — entra no resumo.
public struct StintLap: Sendable, Equatable {
    public let id: String?
    public let tempoMs: Double
    public let valida: Bool

    public init(id: String? = nil, tempoMs: Double, valida: Bool = true) {
        self.id = id; self.tempoMs = tempoMs; self.valida = valida
    }
}

/// Stint mínimo pro resumo. Caller traz `lapsExecutados` se houver
/// telemetria; senão preenche `voltasPlanejadas` (manual).
public struct StintInput: Sendable, Equatable {
    public let id: String?
    public let lapsExecutados: [StintLap]?
    public let voltasPlanejadas: Int?

    public init(id: String? = nil, lapsExecutados: [StintLap]? = nil, voltasPlanejadas: Int? = nil) {
        self.id = id
        self.lapsExecutados = lapsExecutados
        self.voltasPlanejadas = voltasPlanejadas
    }
}

public struct StintResumo: Sendable, Equatable {
    public let voltas: Int
    public let voltasTotais: Int
    public let melhorMs: Double?
    public let melhorLapId: String?
    public let quality: ResumoQuality

    public init(voltas: Int, voltasTotais: Int, melhorMs: Double?, melhorLapId: String?, quality: ResumoQuality) {
        self.voltas = voltas; self.voltasTotais = voltasTotais
        self.melhorMs = melhorMs; self.melhorLapId = melhorLapId
        self.quality = quality
    }
}

public struct DiaResumo: Sendable, Equatable {
    public let voltas: Int
    public let melhorMs: Double?
    public let melhorLapId: String?
    public let melhorStintId: String?
    public let quality: ResumoQuality
    public let calculadoEm: Double          // ms epoch
}

public struct EventoDiaResumo: Sendable, Equatable {
    public let id: String
    public let data: String?               // ISO YYYY-MM-DD

    public init(id: String, data: String? = nil) {
        self.id = id; self.data = data
    }
}

public struct EventoResumoInput: Sendable, Equatable {
    public let dias: [EventoDiaResumo]
    public let stintsByDia: [String: [StintInput]]

    public init(dias: [EventoDiaResumo], stintsByDia: [String: [StintInput]]) {
        self.dias = dias; self.stintsByDia = stintsByDia
    }
}

public struct EventoResumoResult: Sendable, Equatable {
    public let voltasTotal: Int
    public let melhorMsGlobal: Double?
    public let melhorMsPorDia: [String: Double?]
    public let quality: ResumoQuality
    public let calculadoEm: Double
}

public enum EventoResumo {
    /// Resumo de 1 stint.
    public static func calcularStintResumo(_ stint: StintInput?) -> StintResumo {
        guard let stint else { return vazioStint() }
        if let laps = stint.lapsExecutados, !laps.isEmpty {
            let validas = laps.filter { $0.valida && $0.tempoMs.isFinite && $0.tempoMs > 0 }
            let best = melhorDosLaps(validas)
            return StintResumo(
                voltas: validas.count,
                voltasTotais: laps.count,
                melhorMs: best.tempoMs,
                melhorLapId: best.lapId,
                quality: .medido
            )
        }
        if let v = stint.voltasPlanejadas, v > 0 {
            return StintResumo(
                voltas: v, voltasTotais: v,
                melhorMs: nil, melhorLapId: nil,
                quality: .estimado
            )
        }
        return vazioStint()
    }

    /// Resumo de um dia (lista de stints).
    public static func calcularDiaResumo(_ stints: [StintInput], now: Double = Date().timeIntervalSince1970 * 1000) -> DiaResumo {
        guard !stints.isEmpty else {
            return DiaResumo(
                voltas: 0, melhorMs: nil, melhorLapId: nil, melhorStintId: nil,
                quality: .vazio, calculadoEm: now
            )
        }
        var voltas = 0
        var melhorMs: Double? = nil
        var melhorLapId: String? = nil
        var melhorStintId: String? = nil
        var temMedido = false
        var temEstimado = false
        for s in stints {
            let r = calcularStintResumo(s)
            voltas += r.voltas
            if let m = r.melhorMs, melhorMs == nil || m < (melhorMs ?? .infinity) {
                melhorMs = m
                melhorLapId = r.melhorLapId
                melhorStintId = s.id
            }
            switch r.quality {
            case .medido:   temMedido = true
            case .estimado: temEstimado = true
            default: break
            }
        }
        let quality: ResumoQuality
        if temMedido && temEstimado    { quality = .parcial }
        else if temMedido              { quality = .medido }
        else if temEstimado            { quality = .estimado }
        else                           { quality = .vazio }
        return DiaResumo(
            voltas: voltas,
            melhorMs: melhorMs,
            melhorLapId: melhorLapId,
            melhorStintId: melhorStintId,
            quality: quality,
            calculadoEm: now
        )
    }

    /// Resumo de evento (lista de dias × stintsByDia).
    public static func calcularEventoResumo(
        _ evento: EventoResumoInput,
        now: Double = Date().timeIntervalSince1970 * 1000
    ) -> EventoResumoResult {
        guard !evento.dias.isEmpty else {
            return EventoResumoResult(
                voltasTotal: 0, melhorMsGlobal: nil,
                melhorMsPorDia: [:], quality: .vazio, calculadoEm: now
            )
        }
        var voltasTotal = 0
        var melhorMsGlobal: Double? = nil
        var melhorMsPorDia: [String: Double?] = [:]
        var qualidades: [ResumoQuality] = []
        for d in evento.dias {
            let stints = evento.stintsByDia[d.id] ?? []
            let dr = calcularDiaResumo(stints, now: now)
            voltasTotal += dr.voltas
            melhorMsPorDia.updateValue(dr.melhorMs, forKey: d.id)
            if let m = dr.melhorMs, melhorMsGlobal == nil || m < (melhorMsGlobal ?? .infinity) {
                melhorMsGlobal = m
            }
            qualidades.append(dr.quality)
        }
        let naoVazias = qualidades.filter { $0 != .vazio }
        let quality: ResumoQuality
        if naoVazias.isEmpty                        { quality = .vazio }
        else if naoVazias.allSatisfy({ $0 == .medido })   { quality = .medido }
        else if naoVazias.allSatisfy({ $0 == .estimado }) { quality = .estimado }
        else                                        { quality = .parcial }
        return EventoResumoResult(
            voltasTotal: voltasTotal,
            melhorMsGlobal: melhorMsGlobal,
            melhorMsPorDia: melhorMsPorDia,
            quality: quality,
            calculadoEm: now
        )
    }
}

// MARK: - Helpers

private func vazioStint() -> StintResumo {
    StintResumo(
        voltas: 0, voltasTotais: 0,
        melhorMs: nil, melhorLapId: nil,
        quality: .vazio
    )
}

private func melhorDosLaps(_ laps: [StintLap]) -> (tempoMs: Double?, lapId: String?) {
    var best: Double? = nil
    var bestId: String? = nil
    for l in laps {
        guard l.valida, l.tempoMs.isFinite, l.tempoMs > 0 else { continue }
        if best == nil || l.tempoMs < (best ?? .infinity) {
            best = l.tempoMs
            bestId = l.id
        }
    }
    return (best, bestId)
}
